### PR TITLE
Adding custom header in order to provide a POST with custom MIME types

### PR DESCRIPTION
**Pull request by ghost**
Our server desires a oAuth-request with a POST MIME type of 'application/x-www-form-urlencoded'. In order to get it work with the OAuth2Client we needed to add the support of creating the desired POST request.

The OAuth2Client in this state will only work with application/x-www-form-urlencoded and the 'multipart/form-data'. More MIME types must be added if needed. But IMO this is a good start to do so.

If the (new) kNXOAuth2AccountStoreConfigurationCustomHeaderFields key is set in the NXOAuth2AccountStore configuration the all header fields of the resulting HTTP-Request are replaced with the value for this key.

### CODE REVIEW

**Comment by ghvillasboas** [comment]
Hi Andreas!
The funny thing is that I was just working on a similar PR for this feature. Also for an app that interacts with Meetup.com. You were faster. :)

My only suggestion is to add your new key (kNXOAuth2AccountStoreConfigurationCustomHeaderFields) to the NXOAuth2AccountStore h/m files. So it's easier to define it when configuring the shared store.

And authors, please accept this PR!

Best.
George.

**Comment by ghost** [comment]
Hi George,

thanks for pointing that out. I just missed to commit the constants... ups ;-)

Best
Andreas

**Comment by ghvillasboas** [comment]
@AndreasTextunes there are some problems with your code. Specially with the external browser flow.
Do you mind if I send a new PR mixing your code and mine?

**Comment by ghost** [comment]
@ghvillasboas If it works for both of us sure.

File: Sources/OAuth2Client/NXOAuth2AccountStore.m [hunk 1]
```diff
@@ -606,7 +611,7 @@ + (NSDictionary *)accountsFromDefaultKeychain;
     result = (__bridge_transfer NSDictionary *)cfResult;
 
     if (status != noErr) {
-        NSAssert1(status == errSecItemNotFound, @"Unexpected error while fetching accounts from keychain: %d", (int)status);
+        NSAssert1(status == errSecItemNotFound, @"Unexpected error while fetching accounts from keychain: %ld", status);
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m]: This actually adds a warning when compiled in 32bit. Either leave the cast to `int` in or use `"%zd"` which adapts to the native architecture word width.

File: Sources/OAuth2Client/NXOAuth2AccountStore.m [hunk 2]
```diff
@@ -627,7 +632,7 @@ + (void)storeAccountsInDefaultKeychain:(NSDictionary *)accounts;
                            data, kSecAttrGeneric,
                            nil];
     OSStatus __attribute__((unused)) err = SecItemAdd((__bridge CFDictionaryRef)query, NULL);
-    NSAssert1(err == noErr, @"Error while adding token to keychain: %d", (int)err);
+    NSAssert1(err == noErr, @"Error while adding token to keychain: %ld", err);
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m]: See above

File: Sources/OAuth2Client/NXOAuth2AccountStore.m [hunk 3]
```diff
@@ -638,7 +643,7 @@ + (void)removeFromDefaultKeychain;
                            serviceName, kSecAttrService,
                            nil];
     OSStatus __attribute__((unused)) err = SecItemDelete((__bridge CFDictionaryRef)query);
-    NSAssert1((err == noErr || err == errSecItemNotFound), @"Error while deleting token from keychain: ld", (int)err);
+    NSAssert1((err == noErr || err == errSecItemNotFound), @"Error while deleting token from keychain: %ld", err);
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2AccountStore.m]: See above

File: Sources/OAuth2Client/NXOAuth2Connection.m [hunk 4]
```diff
@@ -225,18 +225,34 @@ - (void)applyParameters:(NSDictionary *)parameters onRequest:(NSMutableURLReques
...
+            NSInputStream *postBodyStream = [[NXOAuth2PostBodyStream alloc] initWithParameters:parameters];
+            
+            contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@",[(NXOAuth2PostBodyStream *)postBodyStream boundary]];
+            NSString *contentLength = [NSString stringWithFormat:@"%lld", [(NXOAuth2PostBodyStream *)postBodyStream length]];
+            [aRequest setValue:contentType forHTTPHeaderField:@"Content-Type"];
+            [aRequest setValue:contentLength forHTTPHeaderField:@"Content-Length"];
+            
+            [aRequest setHTTPBodyStream:postBodyStream];
+            
+        }else if ([contentType isEqualToString:@"application/x-www-form-urlencoded"]) {
```
Comment (toto) [inline Sources/OAuth2Client/NXOAuth2Connection.m]: Add a space before the else :-)

**Comment by toto** [comment]
This looks very useful indeed, I added a few very tiny tiny complaints.

If you fix those I will merge this in and it will be in the next minor release.

**Comment by ghost** [comment]
Yeah, cool :-) I'll change that complaints you have.

BTW: I didn't changed the NSAsserts... But I'll fix it anyway.

**Comment by toto** [comment]
:+1:

**Pull request merged**